Add new ipaddresses hash for issue #343

wolf31o2 opened the pull request:
- [x] Joyent, OpenStack, Rackspace
- [x] AWS (done in another PR)
- [x] Provisioner Automator
- [x] Server-side support
- [x] UI changes to new addresses
- [x] Cluster templates
- [x] Docs updated

Fixes #343

wolf31o2 committed:
Add new ipaddresses hash for issue #343

wolf31o2 committed:
Update UI to display node['ipaddresses']['access_v4']

wolf31o2 committed:
Updated to use new (not yet implemented) macros

wolf31o2 committed:
CHANGELOG entry

wolf31o2 committed:
Update clusterdefinitions to match new JSON format

wolf31o2 committed:
Move ipaddresses outside result payload

wolf31o2 committed:
Change cookbook to use new bind_v4 address

albertshau committed:
Merge remote-tracking branch 'origin/develop' into feature/ipaddresses

wolf31o2 committed:
Add ipaddr to Gemfile

wolf31o2 committed:
Use bootstrap_ip local variable

wolf31o2 committed:
Use new access_v4 address

dereklwood commented on provisioner/worker/plugins/providers/fog_provider/Gemfile:
```diff
@@ -1,3 +1,4 @@
 source 'https://rubygems.org'
 
 gem 'fog', '~> 1.21.0'
+gem 'ipaddr'
```
I believe ipaddr is stdlib, not a gem:
$ gem install ipaddr
ERROR:  Could not find a valid gem 'ipaddr' (>= 0) in any repository
ERROR:  Possible alternatives: baddr, hwaddr

wolf31o2 replied:
OK. I'll remove that. I am actually poking around on this, at the moment, because I am seeing strange failures with the fog provisioner (all types) and 0.9.7 (yeah, I know)...

wolf31o2 committed:
Don't try to pull in a stdlib gem

wolf31o2 committed:
Explicitly add net/ssh to this plugin, rather than relying on it being loaded somewhere else

wolf31o2 committed:
Merge remote-tracking branch 'origin/develop' into feature/ipaddresses

Conflicts:
	provisioner/worker/plugins/providers/fog_provider/lib/fog_provider/rackspace.rb

albertshau committed:
Use multiple ip addresses in node properties instead of a single
one. Expects a hash of ip type to ip address to come back in
with a successful confirm task.

Also refactoring macro code so it will be easier to add macros
in the future. Each type of basic macro has its own class instead
of having a lot of conditions in a single class.

albertshau committed:
Merge branch 'feature/ipaddresses' of github.com:continuuity/loom into feature/ipaddresses

albertshau committed:
add missing copyright headers.

albertshau committed:
fix integration test.

albertshau commented on provisioner/worker/plugins/providers/fog_provider/lib/fog_provider/joyent.rb:
```diff
@@ -85,11 +85,15 @@ def confirm(inputmap)
 
       # Process results
-      @result['result']['ipaddress'] = bootstrap_ip
+      @result['result']['ipaddress'] = bootstrap_ip # TODO: remove this
```
I took out support for this in the server (in terms of macros), think it's safe to remove this line. Unless it's still being used?

albertshau commented:
provisioner part lgtm

wolf31o2 committed:
Merge remote-tracking branch 'origin/develop' into feature/ipaddresses

Conflicts:
	CHANGELOG.md
	provisioner/worker/plugins/providers/fog_provider/lib/fog_provider/joyent.rb
	provisioner/worker/plugins/providers/fog_provider/lib/fog_provider/openstack.rb
	provisioner/worker/plugins/providers/fog_provider/lib/fog_provider/rackspace.rb

wolf31o2 committed:
Remove legacy result['result']['ipaddress'], per @albertshau

wolf31o2 commented:
:+1: on the Macro re-write! I like it.

albertshau committed:
Merge remote-tracking branch 'origin/develop' into feature/ipaddresses

wolf31o2 committed:
Fix IP output in errors

wolf31o2 committed:
Fix output to use proper variable

albertshau committed:
Merge branch 'feature/ipaddresses' of github.com:continuuity/loom into feature/ipaddresses

albertshau committed:
make ip macros that dont specify an ip type default to the
bind_v4 ip type.

poornachandra commented on server/src/main/java/com/continuuity/loom/cluster/NodeProperties.java:
```diff
@@ -59,12 +62,25 @@ private NodeProperties(String hostname, String ipaddress, int nodenum, String ha
-  public String getIpaddress() {
-    return ipaddress;
+  public Map<String, String> getIPAddresses() {
+    return ipaddresses;
```
It would be to return unmodifiable map here.

poornachandra commented on server/src/main/java/com/continuuity/loom/cluster/NodeProperties.java:
```diff
@@ -44,9 +46,10 @@
+  private NodeProperties(String hostname, Map<String, String> ipaddresses, int nodenum, String hardwaretype,
+                         String imagetype, String flavor, String image, String sshUser,
+                         Set<String> automators, Set<String> services) {
+    this.ipaddresses = ipaddresses;
```
Defensive copy?

poornachandra commented on server/src/main/java/com/continuuity/loom/cluster/NodeProperties.java:
```diff
@@ -150,12 +166,12 @@ public String getSshUser() {
-    this.ipaddress = ipaddress;
+  public void setIpaddresses(Map<String, String> ipaddresses) {
+    this.ipaddresses = ipaddresses;
     ImmutableMap.builder();
```
Unused code.

albertshau committed:
Merge remote-tracking branch 'origin/develop' into feature/ipaddresses

Conflicts:
	server/src/main/java/com/continuuity/loom/cluster/NodeProperties.java

poornachandra commented on server/src/main/java/com/continuuity/loom/codec/json/current/TaskConfigCodec.java:
```diff
@@ -82,7 +82,8 @@ public TaskConfig deserialize(JsonElement json, Type type, JsonDeserializationCo
     // build node properties
     String hostname = context.deserialize(jsonObj.remove("hostname"), String.class);
-    String ipaddress = context.deserialize(jsonObj.remove("ipaddress"), String.class);
+    Map<String, String> ipaddresses = context.deserialize(jsonObj.remove("ipaddresses"),
```
I'm not sure I understand why we get ipaddresses from jsonObj

poornachandra replied:
If this method is only used in test, then it is better to add a comment reflecting that.

poornachandra commented on server/src/main/java/com/continuuity/loom/http/request/FinishTaskRequest.java:
```diff
@@ -61,6 +66,10 @@ public int getStatus() {
   }
 
+  public Map<String, String> getIpaddresses() {
+    return ipaddresses;
```
unmodifiable map?

poornachandra commented on server/src/main/java/com/continuuity/loom/http/request/FinishTaskRequest.java:
```diff
@@ -30,6 +34,7 @@ public FinishTaskRequest(String workerId, String provisionerId, String tenantId,
     this.stdout = stdout;
     this.stderr = stderr;
     this.status = status;
+    this.ipaddresses = ipaddresses;
```
defensive copy?

albertshau committed:
making defensive copies of collections.

poornachandra commented on server/src/main/java/com/continuuity/loom/macro/eval/HostServiceEvaluator.java:
```diff
@@ -0,0 +1,82 @@
+          if (serviceName.equals(service.getName())) {
+            String hostname = clusterNode.getProperties().getHostname();
+            if (hostname == null) {
+              throw new IncompleteClusterException("node has no hostname for macro expansion.");
```
It would help debugging if the node's name is part of the exception message.

poornachandra commented on server/src/main/java/com/continuuity/loom/macro/eval/IPSelfEvaluator.java:
```diff
@@ -0,0 +1,64 @@
+  public List<String> evaluate(Cluster cluster, Set<Node> clusterNodes, Node node) throws IncompleteClusterException {
+    String ip = node.getProperties().getIPAddress(ipType);
+    if (ip == null) {
+      throw new IncompleteClusterException("node has no ip for macro expansion.");
```
It would help debugging if the node's name is part of the exception message.

poornachandra commented on server/src/main/java/com/continuuity/loom/macro/eval/IPServiceEvaluator.java:
```diff
@@ -0,0 +1,85 @@
+          if (serviceName.equals(service.getName())) {
+            String ip = clusterNode.getProperties().getIPAddress(ipType);
+            if (ip == null) {
+              throw new IncompleteClusterException("node has no ip for macro expansion.");
```
It would help debugging if the node's name is part of the exception message.

poornachandra commented on server/src/main/java/com/continuuity/loom/macro/eval/ClusterOwnerEvaluator.java:
```diff
@@ -0,0 +1,36 @@
+
+  @Override
+  public List<String> evaluate(Cluster cluster, Set<Node> clusterNodes, Node node) throws IncompleteClusterException {
+    return Lists.newArrayList(cluster.getAccount().getUserId());
```
This can be a ImmutableList.of(), right?

poornachandra commented on server/src/main/java/com/continuuity/loom/macro/eval/HostSelfEvaluator.java:
```diff
@@ -0,0 +1,39 @@
+    if (hostname == null) {
+      return null;
+    }
+    return Lists.newArrayList(node.getProperties().getHostname());
```
This can be a ImmutableList.of(), right?

poornachandra commented on server/src/main/java/com/continuuity/loom/macro/eval/IPSelfEvaluator.java:
```diff
@@ -0,0 +1,64 @@
+    if (ip == null) {
+      throw new IncompleteClusterException("node has no ip for macro expansion.");
+    }
+    return Lists.newArrayList(ip);
```
This can be a ImmutableList.of(), right?

poornachandra commented on server/src/main/java/com/continuuity/loom/macro/eval/ServiceCardinalityEvaluator.java:
```diff
@@ -0,0 +1,69 @@
+        }
+      }
+    }
+    return Lists.newArrayList(String.valueOf(count));
```
This can be a ImmutableList.of(), right?

poornachandra commented on server/src/main/java/com/continuuity/loom/macro/eval/ServiceInstanceEvaluator.java:
```diff
@@ -0,0 +1,45 @@
+    if (instanceNum < 0) {
+      return null;
+    }
+    return Lists.newArrayList(String.valueOf(instanceNum));
```
This can be a ImmutableList.of(), right?

poornachandra commented:
I'm done with the review.
